### ISSUE TITLE
Questions about the detailed padding behaviour of the neural network

### ISSUE THREAD

**Issue by Karbo123** [NVlabs/tiny-cuda-nn#111]
I really appreciate this amazing work. I recently try to read the cuda source codes of instant-ngp, and know that the instant-ngp uses this tiny-cuda-nn to build encoding modules and highly-optimized neural networks. I'm a bit confused when reading the codes because I can't figure out the behaviour of the used neural networks regarding to padding.

It seems that due to performance reason, the maxtrix multiplication needs to be `aligned`. For example, let's say we have a neural network with `input dimension 13`, and `output dimension 18`, and we use `FullyFusedMLP` (i.e. `alignment is 16`), and to compute correctly, we have to allocate both the padded input data and the padded output data, and put them in the padded buffers. Let's say the batch size is 100, and we put the input data of shape `(13, 100)` (each column is one batch) to a padded matrix of shape `(16, 100)`, where the first 13 rows are stored with our input data; and we provide another padded matrix `B_pad` of shape `(32, 100)`, and we expect the `FullyFusedMLP` will store its forward result to the matrix `B_pad`, where the first 18 rows are the results.

My questions are as follows:
- Will the network write data to the irrelevant padded region? In the above example, the first 18 rows of the padded output matrix `B_pad` are actually what we concern, and the last 14 rows (i.e. 32 - 18) are irrelevant. The network will write some values (e.g. zeros) to these irrelevant regions, or just keep the regions untouched? It's important since the network only processes data dimension which is pow of 2, and the irrelevant regions may also affect the result because they may also be provided as the input to another MLP.
- The input needn't to be aligned, but the output must be aligned, is that right?

Thank you very much!

**Comment by Tom94** [comment]
Hi there, short answer first: if you use `NetworkWithInputEncoding` and stick to `trainer->step()` / `model->inference()`, then all padding is automatically handled for you. You don't need to worry about it. Same if you're using the PyTorch bindings.

More detailed answer below:
1. __tiny-cuda-nn__ will write into the padded regions, but downstream components of __tiny-cuda-nn__ (such as `Loss`) will automatically ignore the padded entries, so it does not matter what is written there.
2. If you are either using the PyTorch bindings are calling `DifferentiableObject::inference()`, then the output must not be aligned. The padding will automatically be done behind the scenes and subsequently sliced away. The rest of the API (e.g. `inference_mixed_precision`, `forward` and `backward`) assumes padded outputs.
3. Inputs must also be aligned (unless you use the PyTorch bindings), but if you use `NetworkWithInputEncoding` you don't need to worry about it: the encodings don't require input alignment and will automatically align their outputs to match the following neural network.

**Comment by Karbo123** [comment]
Thanks for the reply! Actually, I want to build my own network which has 3 neural networks, and 3 encoding modules, so I guess it is better to build the architecture by reimplementing `inference_mixed_precision`, `forward` and `backward` to achieve maximum performance. I find [ngp's nerf_network.h](https://github.com/Karbo123/instant_nerf/blob/master/model/include/neural-graphics-primitives/nerf_network.h) seems to be a great example to learn from, but there are still some problems that are quite hard for me. 

**I summarize serveral guesses which might not be true, and hope you can help me check their correctness, and correct them if they are wrong.** Thank you!

Hypotheses/Guesses (only show what I concern the most):
- for `inference_mixed_precision()` or `forward()`:
  - for `GridEncoding`, padded region will be internally filled with zeros, as shown in [here](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/include/tiny-cuda-nn/encodings/grid.h#L946-L957)
  - for `IdentityEncoding`, padded region will be internally  filled with ones, as shown in [here](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/include/tiny-cuda-nn/encodings/identity.h#L64-L68)
  - for `FullyFusedMLP`, 
    - output (i.e. `m_padded_output_width`) will be internally aligned by 16 although the provided `output_width` might not be aligned, as shown in [here](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/src/fully_fused_mlp.cu#L658)
    - internal width of computation is `m_input_width -> [m_network_width] * N -> m_padded_output_width`, as shown in [here](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/src/fully_fused_mlp.cu#L661-L679)
    - it seems that all the output neurons come from computed results. for example, if we set `output_width=17`, then the network will actually output `m_padded_output_width=32` neurons because the final layer of the network has 32 neurons. all of  the output positions store valid results, padded regions are not filled with a constant value (i.e. not filled with zeros or ones).
    - I still don't know whether the provoded [input_width](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/src/fully_fused_mlp.cu#L637) need to aligned (this may contradict with the statement that `Inputs must also be aligned` made in the third point in your [comment](https://github.com/NVlabs/tiny-cuda-nn/issues/111#issuecomment-1160033548)), because there is no check for the alignment, and it is directly used to allocate the weight matrices, as shown in [here](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/src/fully_fused_mlp.cu#L661-L665).
- for `backward()`:
  - for `FullyFusedMLP`:
    - if `m_output_width != m_padded_output_width`, then the incoming gradient (i.e. [dL_doutput](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/src/fully_fused_mlp.cu#L779)) in the padded region must be filled with zeros.
    - it seems that all the inputs are involved in the computation, so maybe [dL_dinput](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/src/fully_fused_mlp.cu#L780) writes to the buffer which has the same shape as [input](https://github.com/NVlabs/tiny-cuda-nn/blob/b7911744e3/src/fully_fused_mlp.cu#L777) .

I'll very appreciate your explanation, thanks!

**Comment by Tom94** [comment]
These observations are all correct!

To elaborate on input alignment: inputs to `FullyFusedMLP` need to indeed be multiples of 16 wide -- and you're right that it's unsafe that the existing code does not explicitly check this. It should!
